proccontrol/src/process.C

```diff
@@ -242,13 +244,56 @@ bool int_process::attachThreads()
          }
          pthrd_printf("Creating new thread for %d/%d during attach\n", pid, *i);
          thr = int_thread::createThread(this, NULL_THR_ID, *i, false, int_thread::as_needs_attach);
-         found_new_threads = true;
+         found_new_threads = loop_new_threads = true;
       }
-   } while (found_new_threads);
+   } while (loop_new_threads);
 
    return true;
 }
 
+bool int_process::attachThreads()
+{
+   bool found_new_threads = false;
+   return attachThreads(found_new_threads);
+}
+
+// Attach any new threads and synchronize, until there are no new threads
+bool int_process::attachThreadsSync()
+{
+   while (true) {
+      bool found_new_threads = false;
+
+      ProcPool()->condvar()->lock();
+      bool result = attachThreads(found_new_threads);
+      if (found_new_threads)
+         ProcPool()->condvar()->broadcast();
+      ProcPool()->condvar()->unlock();
+
+      if (!result) {
+         pthrd_printf("Failed to attach to threads in %d\n", pid);
+         setLastError(err_internal, "Could not get threads during attach\n");
+         return false;
+      }
+
+      if (!found_new_threads)
+         return true;
+
+      pthrd_printf("Wait again for attach from process %d\n", pid);
+      bool proc_exited = false;
+      result = waitAndHandleForProc(true, this, proc_exited);
```

Comment:
Thanks for working on this, Josh. A few points: 

- This is waiting for one event to happen on any thread in the process.  It's not specifically waiting for an attach event and it's not waiting for an event from every thread.  I think we could proceed out of this loop with things still in the half-attached state.  I suspect this is more narrowing the possibility of the race rather than fixing it.

- This is unconditionally waiting for events.  I'm not sure whether every platform and situation's attachThreads() generates an event that can be waited for (maybe they all do, I'm honestly not sure).  The prior code only waited for events if the int_process's getState() wasn't 'running'.

- Finally, this is going to cause performance problems on BG/Q.  We were previously using group operations to structure attach like "trigger attach to X number of process"  and then "wait for attach on X number of processes".  That worked well with BGQ's high-latency asynchronous debug interface.  This is moving to a "attach to process 1, wait for process 1, attach to process 2, ...".

For the first item, I think you could block with loop waiting for all threads reaching neonatal state.  The second two items could be fixed by moving this code to be linux specific.